restic prune: `Fatal: index is not complete`

Output of `restic version`
--------------------------

`restic 0.14.0-dev (compiled manually) compiled with go1.18 on linux/amd64`

How did you run restic exactly?
-------------------------------

```
$ restic --cacert /path/to/file.pem -v prune --max-unused 0
repository f72c6108 opened (version 2, compression level auto)
loading indexes...
loading all snapshots...
finding data that is still in use for 18 snapshots
[11:26] 100.00%  18 / 18 snapshots
searching used packs...
{<data/[null]> <data/00000000>} not found in the index

Integrity check failed: Data seems to be missing.
Will not start prune to prevent (additional) data loss!
Please report this error (along with the output of the 'prune' run) at
https://github.com/restic/restic/issues/new/choose
Fatal: index is not complete
```

Environment variables:

```
RESTIC_REPOSITORY=s3:https://<redacted>
RESTIC_PASSWORD=<redacted>
AWS_ACCESS_KEY_ID=<redacted>
AWS_SECRET_ACCESS_KEY=<redacted>
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

S3


Expected behavior
-----------------

prune finishes successfully

Actual behavior
---------------

prune aborts with an error

Steps to reproduce the behavior
-------------------------------

As of now, I'm not aware on how to consistently reproduce this issue.

Do you have any idea what may have caused this?
-----------------------------------------------

Previously, a backup run was terminated by the oom killer because the machine went out of RAM.

Do you have an idea how to solve the issue?
-------------------------------------------

There are similiar bug reports for this kind of issue, but I'm not sure if the described solutions (e.g. remove affected snapshot) apply here as well - hence this issue. :)

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

restic is just awesome 🥳

ok, now while the first backup run succeeded, this one failed:

```
open repository
lock repository
using parent snapshot e80302d1
load index files
start scan on [/ /foo /bar]
start backup on [/ /foo /bar]
scan finished in 11090.200s: 18448691 files, 3.583 TiB
panic: completed file with null ID
 
goroutine 144 [running]:
github.com/restic/restic/internal/archiver.(*FileSaver).saveFile.func1()
	/restic/internal/archiver/file_saver.go:133 +0x2ad
github.com/restic/restic/internal/archiver.(*FileSaver).saveFile.func3({{0x85, 0xc7, 0x7, 0x13, 0xe3, 0x84, 0x39, 0x4b, 0x70, 0xc3, ...}, ...})
	/restic/internal/archiver/file_saver.go:215 +0xe9
github.com/restic/restic/internal/archiver.(*BlobSaver).worker(0x532f3c3037333533?, {0x11fcb50, 0xc000132cc0}, 0xc0451300c0)
	/restic/internal/archiver/blob_saver.go:100 +0x1ed
github.com/restic/restic/internal/archiver.NewBlobSaver.func1()
	/restic/internal/archiver/blob_saver.go:33 +0x29
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20220819030929-7fc1605a5dde/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
	/home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20220819030929-7fc1605a5dde/errgroup/errgroup.go:72 +0xa5
```

another backup run, slightly different panic:

```
open repository
lock repository
using parent snapshot e80302d1
load index files
start scan on [/ /foo /bar]
start backup on [/ /foo /bar]
error: lstat /proc/1405818: no such file or directory
error: lstat /proc/1406250: no such file or directory
error: lstat /proc/1406261: no such file or directory
scan finished in 8489.563s: 18413327 files, 3.522 TiB
error: lstat <redacted>: no such file or directory
panic: completed file with null ID

goroutine 110 [running]:
github.com/restic/restic/internal/archiver.(*FileSaver).saveFile.func1()
        /restic/internal/archiver/file_saver.go:133 +0x2ad
github.com/restic/restic/internal/archiver.(*FileSaver).saveFile.func3({{0xfc, 0x4, 0x82, 0xe3, 0x68, 0xb1, 0xa, 0xa4, 0xc8, 0x19, ...}, ...})
        /restic/internal/archiver/file_saver.go:215 +0xe9
github.com/restic/restic/internal/archiver.(*BlobSaver).worker(0xc7bda32d5d2ec912?, {0x11fcb50, 0xc000ec42c0}, 0xc051aed3e0)
        /restic/internal/archiver/blob_saver.go:100 +0x1ed
github.com/restic/restic/internal/archiver.NewBlobSaver.func1()
        /restic/internal/archiver/blob_saver.go:33 +0x29
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20220819030929-7fc1605a5dde/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        /home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20220819030929-7fc1605a5dde/errgroup/errgroup.go:72 +0xa5
```

--- a/internal/archiver/blob_saver_test.go
+++ b/internal/archiver/blob_saver_test.go
@@ -48,14 +48,19 @@ func TestBlobSaver(t *testing.T) {
 
 	var wait sync.WaitGroup
 	var results []SaveBlobResponse
+	var lock sync.Mutex
 
 	wait.Add(20)
 	for i := 0; i < 20; i++ {
 		buf := &Buffer{Data: []byte(fmt.Sprintf("foo%d", i))}
 		idx := i
+		lock.Lock()
 		results = append(results, SaveBlobResponse{})
+		lock.Unlock()
 		b.Save(ctx, restic.DataBlob, buf, func(res SaveBlobResponse) {
+			lock.Lock()
 			results[idx] = res
+			lock.Unlock()
 			wait.Done()
 		})
 	}

--- a/internal/archiver/file_saver.go
+++ b/internal/archiver/file_saver.go
@@ -199,7 +199,10 @@ func (s *FileSaver) saveFile(ctx context.Context, chnker *chunker.Chunker, snPat
 
 		// add a place to store the saveBlob result
 		pos := idx
+
+		lock.Lock()
 		node.Content = append(node.Content, restic.ID{})
+		lock.Unlock()
 
 		s.saveBlob(ctx, restic.DataBlob, buf, func(sbr SaveBlobResponse) {
 			lock.Lock()

